dev-docs/RFCs/v8.0/luma-refactor.md

```diff
@@ -0,0 +1,98 @@
+# RFC: Refactoring luma.gl
+
+* **Author**: Tarek Sherif
+* **Date**: September, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC proposes steps to simplify luma.gl's API and ease its adoption among users.
+
+
+## Background
+
+luma.gl is currently a very large library relative to the functionality it provides and is difficult to approach for new users. This is due to several factors, among them the following:
+- 14 separate modules with no clear sense of how they're meant to interoperate
+- 201 exported classes and functions, again without a clear sense of their relationships
+- A large code base. The minified dist files (dist.min.js) total 981kB with the webgl re-exports from core removed. Compare to the minified build files of e.g. a full 3D engine like three.js at 592kB, a low-level library like PicoGL at 66kB. Breakdown of build file sizes for luma.gl modules is as follows:
+
+| Module      | Size |
+| ----------- | ----------- |
+|core|235kB|
+|debug|202kB|
+|webgl|145kB|
+|addons|134kB|
+|shadertools|97kB|
+|effects|72kB|
+|webgl-state-tracker|30kB|
+|gpgpu|25kB|
+|webgl2-polyfill|21kB|
+|constants|16kB|
+|**total**|**981kB**|
+
+- Several unfinished or unused APIs that make it hard to tell what's reliable
+- Documentation is sprawling and not focused on the API. Examples demonstrate random scenes rather than API features
+
+Clearly defining luma.gl's goals as a framework, and structuring the codebase and documentation around those goals, would benefit it in the following ways:
+- Reduce its complexity and associated maintenance costs
+- Reduce the weight it imposes on applications that use it
+- Make its value proposition clear to potential users
+- Simplify adoption
+
+## Overview
+
+The first step towards simplifying luma.gl is to be explicit about what its goals are. I propose that changes be made around the following key principles:
```

Comment:
@tsherif , another aspect that should be considered for discussion is , auto deduction unspecified parameters, this makes easy for the users, but does create complexity in the code. (Search for `deduce` in code base for references)

Reply:
Very good point. I'll add some text about that.